File: crates/algokit_utils/src/clients/algorand_client.rs
use crate::clients::app_manager::AppManager;
use crate::clients::asset_manager::AssetManager;
use crate::clients::client_manager::ClientManager;
use crate::clients::network_client::{AlgoConfig, AlgorandService};
use crate::transactions::{
    Composer, ComposerParams, TransactionComposerConfig, TransactionCreator, TransactionSender,
};
use crate::{AccountManager, TransactionSigner};
use algod_client::models::TransactionParams;
use algokit_transact::Address;
use std::sync::{Arc, Mutex};

pub struct AlgorandClient {
    client_manager: Arc<ClientManager>,
    asset_manager: Arc<AssetManager>,
    app_manager: Arc<AppManager>,
    transaction_sender: Arc<TransactionSender>,
    transaction_creator: Arc<TransactionCreator>,
    account_manager: Arc<Mutex<AccountManager>>,
    default_composer_config: Option<TransactionComposerConfig>,
}

/// A client that brokers easy access to Algorand functionality.
pub struct AlgorandClientParams {
    pub client_config: AlgoConfig,
    pub composer_config: Option<TransactionComposerConfig>,
}

impl AlgorandClient {
    pub fn new(params: &AlgorandClientParams) -> Self {
        let client_manager = ClientManager::new(&params.client_config).unwrap();
        let algod_client = client_manager.algod();

        let account_manager = Arc::new(Mutex::new(AccountManager::new()));

        // TODO: an easy way to create algorand client without this newGroup
        // TODO: convert this to trait
        let new_group = {
            let algod_client = algod_client.clone();
            let account_manager = account_manager.clone();
            let default_composer_config = params.composer_config.clone();
            move |composer_config: Option<TransactionComposerConfig>| {
                Composer::new(ComposerParams {
                    algod_client: algod_client.clone(),
                    signer_getter: account_manager.clone(),
                    composer_config: composer_config.or_else(|| default_composer_config.clone()),
                })
            }
        };

        let asset_manager = Arc::new(AssetManager::new(algod_client.clone(), new_group.clone()));
        let app_manager = Arc::new(AppManager::new(algod_client.clone()));

        // Create closure for new_group function
        let transaction_sender = TransactionSender::new(
            new_group.clone(),
            asset_manager.clone(),
            app_manager.clone(),
        );

        // Create closure for TransactionCreator
        let transaction_creator = TransactionCreator::new(new_group.clone());

        Self {
            client_manager: client_manager.into(),
            account_manager: account_manager.clone(),
            asset_manager: asset_manager.clone(),
            app_manager: app_manager.clone(),
            transaction_sender: transaction_sender.into(),
            transaction_creator: transaction_creator.into(),
            default_composer_config: params.composer_config.clone(),
        }
    }

    pub async fn get_suggested_params(
        &self,
    ) -> Result<TransactionParams, Box<dyn std::error::Error>> {
        Ok(self.client_manager.algod().transaction_params().await?)
    }

    pub fn client(&self) -> Arc<ClientManager> {
        self.client_manager.clone()
    }

    /// Get access to the AssetManager for asset operations
    pub fn asset(&self) -> Arc<AssetManager> {
        self.asset_manager.clone()
    }

    /// Get access to the AppManager for app operations
    pub fn app(&self) -> Arc<AppManager> {
        self.app_manager.clone()
    }

    /// Get access to the TransactionSender for sending transactions
    pub fn send(&self) -> Arc<TransactionSender> {
        self.transaction_sender.clone()
    }

    /// Get access to the TransactionCreator for building transactions
    pub fn create(&self) -> Arc<TransactionCreator> {
        self.transaction_creator.clone()
    }

    pub fn account(&self) -> Arc<Mutex<AccountManager>> {
        self.account_manager.clone()
    }

    /// Create a new transaction composer for building transaction groups
    pub fn new_group(&self, params: Option<TransactionComposerConfig>) -> Composer {
        Composer::new(ComposerParams {
            algod_client: self.client_manager.algod().clone(),
            signer_getter: self.account_manager.clone(),
            composer_config: params.or_else(|| self.default_composer_config.clone()),
        })
    }

    pub fn default_localnet(params: Option<TransactionComposerConfig>) -> Self {
        Self::new(&AlgorandClientParams {
            client_config: AlgoConfig {
                algod_config: ClientManager::get_default_localnet_config(AlgorandService::Algod),
                indexer_config: Some(ClientManager::get_default_localnet_config(
                    AlgorandService::Indexer,
                )),
                kmd_config: Some(ClientManager::get_default_localnet_config(
                    AlgorandService::Kmd,
                )),
            },
            composer_config: params,
        })
    }

    pub fn testnet(params: Option<TransactionComposerConfig>) -> Self {
        Self::new(&AlgorandClientParams {
            client_config: AlgoConfig {
                algod_config: ClientManager::get_algonode_config("testnet", AlgorandService::Algod),
                indexer_config: Some(ClientManager::get_algonode_config(
                    "testnet",
                    AlgorandService::Indexer,
                )),
                kmd_config: None,
            },
            composer_config: params,
        })
    }

    pub fn mainnet(params: Option<TransactionComposerConfig>) -> Self {
        Self::new(&AlgorandClientParams {
            client_config: AlgoConfig {
                algod_config: ClientManager::get_algonode_config("mainnet", AlgorandService::Algod),
                indexer_config: Some(ClientManager::get_algonode_config(
                    "mainnet",
                    AlgorandService::Indexer,
                )),
                kmd_config: None,
            },
            composer_config: params,
        })
    }

    pub fn from_environment(params: Option<TransactionComposerConfig>) -> Self {
        Self::new(&AlgorandClientParams {
            client_config: ClientManager::get_config_from_environment_or_localnet(),
            composer_config: params,
        })
    }

    pub fn set_signer(&mut self, sender: Address, signer: Arc<dyn TransactionSigner>) {
        self.account_manager
            .lock()
            .unwrap()
            .set_signer(sender, signer);
    }
}

File: packages/typescript/algokit_utils/src/algorand-client.ts
import { AppManager } from './clients/app-manager'
import { AssetManager } from './clients/asset-manager'
import { ClientManager } from './clients/client-manager'
import { AlgoConfig, AlgorandService } from './clients/network-client'
import {
  TransactionComposer,
  TransactionComposerConfig,
  TransactionComposerParams,
} from './transactions/composer'
import { SignerGetter, TransactionSigner } from './transactions/common'
import type { TransactionParams } from '@algorandfoundation/algod-client'

export type AlgorandClientParams =
  | {
      clientConfig: AlgoConfig
      clientManager?: undefined
      composerConfig?: TransactionComposerConfig
    }
  | {
      clientConfig?: undefined
      clientManager: ClientManager
      composerConfig?: TransactionComposerConfig
    }

class SignerRegistry implements SignerGetter {
  private readonly signers = new Map<string, TransactionSigner>()
  private defaultSigner?: TransactionSigner

  setSigner(address: string, signer: TransactionSigner) {
    this.signers.set(address, signer)
  }

  clearSigner(address: string) {
    this.signers.delete(address)
  }

  setDefaultSigner(signer: TransactionSigner) {
    this.defaultSigner = signer
  }

  clearDefaultSigner() {
    this.defaultSigner = undefined
  }

  getSigner(address: string): TransactionSigner {
    const signer = this.signers.get(address) ?? this.defaultSigner
    if (!signer) {
      throw new Error(`No signer registered for address ${address}. Use setSigner or setDefaultSigner to configure one.`)
    }
    return signer
  }
}

/**
 * A client that brokers easy access to Algorand functionality.
 */
export class AlgorandClient {
  private readonly clientManager: ClientManager
  private readonly assetManager: AssetManager
  private readonly appManager: AppManager
  private readonly signerRegistry = new SignerRegistry()
  private readonly defaultComposerConfig?: TransactionComposerConfig

  constructor(params: AlgorandClientParams) {
    if (!params.clientManager && !params.clientConfig) {
      throw new Error('AlgorandClient requires either a client configuration or an existing ClientManager instance.')
    }

    this.clientManager = params.clientManager ?? new ClientManager(params.clientConfig)
    this.defaultComposerConfig = params.composerConfig

    this.assetManager = new AssetManager(this.clientManager.algod, () => this.newComposer())
    this.appManager = new AppManager(this.clientManager.algod)
  }

  /** Creates a new transaction composer pre-configured with the Algorand client context. */
  newComposer(composerConfig?: TransactionComposerConfig): TransactionComposer {
    const params: TransactionComposerParams = {
      algodClient: this.clientManager.algod,
      signerGetter: this.signerRegistry,
      composerConfig: composerConfig ?? this.defaultComposerConfig,
    }
    return new TransactionComposer(params)
  }

  /** Registers a signer for a specific address. */
  setSigner(address: string, signer: TransactionSigner): this {
    this.signerRegistry.setSigner(address, signer)
    return this
  }

  /** Removes a previously registered signer for an address. */
  clearSigner(address: string): this {
    this.signerRegistry.clearSigner(address)
    return this
  }

  /** Registers a default signer used when no address-specific signer exists. */
  setDefaultSigner(signer: TransactionSigner): this {
    this.signerRegistry.setDefaultSigner(signer)
    return this
  }

  /** Clears the default signer. */
  clearDefaultSigner(): this {
    this.signerRegistry.clearDefaultSigner()
    return this
  }

  /** Returns the underlying ClientManager. */
  get client(): ClientManager {
    return this.clientManager
  }

  /** Returns the AssetManager helper. */
  get asset(): AssetManager {
    return this.assetManager
  }

  /** Returns the AppManager helper. */
  get app(): AppManager {
    return this.appManager
  }

  /** Retrieves suggested transaction parameters from algod. */
  async getSuggestedParams(): Promise<TransactionParams> {
    return await this.clientManager.algod.transactionParams()
  }

  /** Creates an AlgorandClient from a raw network configuration. */
  static fromConfig(clientConfig: AlgoConfig, composerConfig?: TransactionComposerConfig): AlgorandClient {
    return new AlgorandClient({ clientConfig, composerConfig })
  }

  /** Creates an AlgorandClient from an existing ClientManager. */
  static fromClientManager(clientManager: ClientManager, composerConfig?: TransactionComposerConfig): AlgorandClient {
    return new AlgorandClient({ clientManager, composerConfig })
  }

  /** Creates an AlgorandClient configured for a local development network. */
  static localnet(composerConfig?: TransactionComposerConfig): AlgorandClient {
    return AlgorandClient.fromConfig({
      algodConfig: ClientManager.getDefaultLocalnetConfig(AlgorandService.Algod),
      indexerConfig: ClientManager.getDefaultLocalnetConfig(AlgorandService.Indexer),
      kmdConfig: ClientManager.getDefaultLocalnetConfig(AlgorandService.Kmd),
    }, composerConfig)
  }

  /** Creates an AlgorandClient configured for Algonode TestNet. */
  static testnet(composerConfig?: TransactionComposerConfig): AlgorandClient {
    return AlgorandClient.fromConfig({
      algodConfig: ClientManager.getAlgonodeConfig('testnet', AlgorandService.Algod),
      indexerConfig: ClientManager.getAlgonodeConfig('testnet', AlgorandService.Indexer),
      kmdConfig: undefined,
    }, composerConfig)
  }

  /** Creates an AlgorandClient configured for Algonode MainNet. */
  static mainnet(composerConfig?: TransactionComposerConfig): AlgorandClient {
    return AlgorandClient.fromConfig({
      algodConfig: ClientManager.getAlgonodeConfig('mainnet', AlgorandService.Algod),
      indexerConfig: ClientManager.getAlgonodeConfig('mainnet', AlgorandService.Indexer),
      kmdConfig: undefined,
    }, composerConfig)
  }

  /** Creates an AlgorandClient from environment configuration or defaults to localnet. */
  static fromEnvironment(composerConfig?: TransactionComposerConfig): AlgorandClient {
    return AlgorandClient.fromConfig(ClientManager.getConfigFromEnvironmentOrLocalNet(), composerConfig)
  }
}
